operator/pkg/ingress/ingress_class_test.go

```diff
@@ -4,96 +4,558 @@
 	"testing"
 
-	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/workqueue"
 
+	"github.com/cilium/cilium/operator/k8s"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
 	slim_networkingv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/networking/v1"
 	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
 )
 
-func Test_ingressClassHandleEvent(t *testing.T) {
+func Test_ingressClassSyncCanExit(t *testing.T) {
+	_, cs := k8sClient.NewFakeClientset()
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
 
-	t.Run("unknown event", func(t *testing.T) {
-		i := ingressClassManager{
-			ingressQueue: newQueue(),
-		}
-		err := i.handleEvent(ingressAddedEvent{})
-		require.Error(t, err)
-	})
+	var ingressClasses resource.Resource[*slim_networkingv1.IngressClass]
 
-	t.Run("delete ingressClass", func(t *testing.T) {
-		i := ingressClassManager{
-			ingressQueue: newQueue(),
-		}
-		err := i.handleEvent(ingressClassDeletedEvent{ingressClass: &slim_networkingv1.IngressClass{
-			ObjectMeta: slim_metav1.ObjectMeta{
-				Name: ciliumIngressClassName,
-			},
-		}})
-		require.NoError(t, err)
-		require.Equal(t, 1, i.ingressQueue.Len())
-	})
-
-	t.Run("add ingressClass", func(t *testing.T) {
-		t.Run("no ops if name is not us", func(t *testing.T) {
-			i := ingressClassManager{
-				ingressQueue: newQueue(),
-			}
-			err := i.handleEvent(ingressClassAddedEvent{ingressClass: &slim_networkingv1.IngressClass{}})
-			require.NoError(t, err)
-			require.Empty(t, i.ingressQueue.Len())
-		})
+	h := hive.New(
+		cell.Provide(func() k8sClient.Clientset { return cs }),
+		cell.Provide(k8s.IngressClassResource),
+		cell.Invoke(func(r resource.Resource[*slim_networkingv1.IngressClass]) {
+			ingressClasses = r
+		}),
+	)
 
-		t.Run("with correct name", func(t *testing.T) {
-			i := ingressClassManager{
-				ingressQueue: newQueue(),
-			}
-			err := i.handleEvent(ingressClassAddedEvent{ingressClass: &slim_networkingv1.IngressClass{
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	if err := h.Start(ctx); err != nil {
+		t.Fatalf("hive.start failed: %s", err)
+	}
+
+	i := newIngressClassManager(ctx, queue, ingressClasses)
+
+	// Start the ingressClassManager
+	go i.Run(ctx)
+
+	// Wait for an initial sync.
+	if err := i.WaitForSync(ctx); err != nil {
+		t.Fatalf("unexpected error while doing initial sync: %s", err)
+	}
+	assert.True(t, i.synced.Load())
+
+	// This second sync should not block.
+	if err := i.WaitForSync(ctx); err != nil {
+		t.Fatalf("unexpected error while doing second sync: %s", err)
+	}
+
+	// Try exiting a sync using context
+	testCtx, testCancel := context.WithCancel(ctx)
+	testCancel()
+	i.synced.Store(false)
+
+	err := i.WaitForSync(testCtx)
```

Comment:
nit: combining the related if statements

```
if err := i.WaitForSync(testCtx); err == nil {
...
} else if !errors.Is(err, context.Canceled) {
...
```

Reply:
Oh sweet, I didn't know that you could reference variables defined in an if statement in other branches.